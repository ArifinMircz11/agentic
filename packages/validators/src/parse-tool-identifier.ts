import type { ParsedToolIdentifier } from './types'
import { parseToolUri } from './parse-tool-uri'

export function parseToolIdentifier(
  identifier: string
  // { namespace }: { namespace?: string } = {}
): ParsedToolIdentifier | undefined {
  if (!identifier) {
    return
  }

  let uri = identifier
  try {
    const { pathname } = new URL(identifier)
    uri = pathname
  } catch {}

  uri = uri.replaceAll(/^\//g, '')
  uri = uri.replaceAll(/\/$/g, '')

  if (!uri.length) {
    return
  }

  // const hasNamespacePrefix = /^([a-zA-Z0-9-]{1,64}\/)/.test(uri)

  // if (!hasNamespacePrefix) {
  //   if (namespace) {
  //     // add inferred namespace prefix (defaults to authenticated user's username)
  //     uri = `${namespace}/${uri}`
  //   } else {
  //     // throw new Error(`FaaS identifier is missing namespace prefix or you must be authenticated [${uri}]`)
  //     return
  //   }
  // }

  return parseToolUri(uri)
}
