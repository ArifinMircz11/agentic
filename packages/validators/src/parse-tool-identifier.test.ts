import { describe, expect, test } from 'vitest'

import { parseToolIdentifier } from './parse-tool-identifier'
import {
  isValidDeploymentHash,
  isValidDeploymentIdentifier,
  isValidNamespace,
  isValidProjectIdentifier,
  isValidProjectSlug,
  isValidToolName
} from './validators'

function success(...args: Parameters<typeof parseToolIdentifier>) {
  const result = parseToolIdentifier(...args)
  expect(result).toBeTruthy()
  expect(result!.projectIdentifier).toBeTruthy()
  expect(result!.projectNamespace).toBeTruthy()
  expect(result!.projectSlug).toBeTruthy()
  expect(result!.deploymentIdentifier).toBeTruthy()
  expect(result!.deploymentVersion || result!.deploymentHash).toBeTruthy()
  expect(isValidProjectIdentifier(result!.projectIdentifier)).toBe(true)
  expect(isValidProjectSlug(result!.projectSlug)).toBe(true)
  expect(isValidNamespace(result!.projectNamespace)).toBe(true)
  expect(isValidDeploymentIdentifier(result!.deploymentIdentifier)).toBe(true)
  expect(isValidToolName(result!.toolName)).toBe(true)

  if (result!.deploymentHash) {
    expect(isValidDeploymentHash(result!.deploymentHash)).toBe(true)
  }

  expect(result).toMatchSnapshot()
}

function error(...args: Parameters<typeof parseToolIdentifier>) {
  expect(() => parseToolIdentifier(...args)).throws()
}

describe('parseToolIdentifier', () => {
  test('strict mode valid', () => {
    success('@username/foo-bar@01234567/foo')
    success('@username/foo-bar@01234567/foo')
    success('@username/foo-bar@abc123lz/foo')
    success('@username/foo-bar/foo')
    success('@username/foobar123-yo@01234567/foo_bar_BAR_901')
    success('@username/foobar@01234567/get_weather01')
    success('@username/foobar@latest/foo')
    success('@username/foobar@dev/foo')
    success('@username/foobar@1.0.0/foo')

    success('@username/foo-bar@latest/foo')
    success('@username/foo-bar@dev/foo')
    success('@username/foo-bar@1.0.0/foo')
    success('@username/foobar123-yo@0.0.1/foo_bar_BAR_901')
    success('@username/foobar123-yo@0.0.1/foo')

    success('@u/foo-bar/foo')
    success('@a/foo-bar/foo_123')
    success('@foo/foobar123-yo/foo_bar_BAR_901')
    success('@foo/foobar123-yo/foo')
    success('@foo/bar/baz')
  })

  test('strict mode invalid', () => {
    error()
    error('')
    error('foo')
    error('foo/bar')
    error('foo/bar/baz')
    error('foo/bar/baz/qux')
    error('@foo/bar/baz/qux')
    error('@foo/bar/baz/qux/quux')
    error('@foo/bar/baz/qux/quux/corge')
    error('@foo/bar/baz/qux/quux/corge/grault')
    error('@foo/bar/baz/qux/quux/corge/grault/garply')
    error('@foo/bar/baz/')
    error('@foo')
    error('@foo/bar')
    error('@foo/bar/')

    error('@foo-bar@01234567/foo')
    error('@%/foo-bar@01234567/foo')
    error('@user/foo^bar@01234567/foo')
    error('@user@foo^bar@01234567/foo')
    error('@username/Foo-Bar@01234567/foo')
    error('username/foo-bar/foo')
    error('@Username/foo-bar/foo')
    error('@username/Foo-bar/foo')
    error('@username/foo-bar/')
    error('@username/foo-bar')

    error('@foo_bar@latest/foo')
    error('@username/foo-bar@1.0.0/foo@')
    error('@username/foo-bar@/foo')
    error('@username/foo-bar@/foo/')
    error('@username/fooBar123-yo@0.0.1/foo/bar/123-456')

    error('//@username/foo-bar@01234567/foo')
    error('https://@username/foo-bar@01234567/foo')
    error('https://example.com/@username/foo-bar/foo')
  })

  test('non-strict mode valid', () => {
    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo', {
      strict: false
    })
    success('/@username/foo-bar@01234567/foo', { strict: false })
    success('@username/foo-bar@01234567/foo/', { strict: false })
    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })

    success('/@username/foo-bar@01234567/foo/', { strict: false })
    success('/@username/foo-bar@latest/foo/', { strict: false })
    success('/@username/foo-bar@dev/foo/', { strict: false })
    success('/@username/foo-bar@1.0.0/foo/', { strict: false })
    success('/@username/foo-bar/foo/', { strict: false })

    success('https://gateway.agentic.so/@username/foo-bar@01234567/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar/foo/', {
      strict: false
    })
    success('https://gateway.agentic.so/@username/foo-bar@latest/foo/', {
      strict: false
    })
  })

  test('non-strict mode invalid', () => {
    error(undefined, { strict: false })
    error('', { strict: false })
    error('https://gateway.agentic.so', { strict: false })
    error('//gateway.agentic.so', { strict: false })
    error('https://gateway.agentic.so/@username', { strict: false })
    error('https://gateway.agentic.so/@username/foo-bar', { strict: false })
    error('https://gateway.agentic.so/call/@username/foo-bar@latest/tool', {
      strict: false
    })
    error('/call/@username/foo-bar/foo', { strict: false })
    error('//@username/foo-bar/foo', { strict: false })
    error('@username/foo-bar/foo//', { strict: false })

    error('@username/foo-bar/fooðŸ˜€', { strict: false })
    error('@username/Foo-Bar@dev/foo/', { strict: false })
  })
})
