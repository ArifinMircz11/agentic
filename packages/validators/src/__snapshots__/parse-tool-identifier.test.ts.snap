// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`URL prefix and suffix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`URL prefix and suffix success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
}
`;

exports[`URL prefix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`URL prefix success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`URL prefix success 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
}
`;

exports[`URL prefix success 6`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
  "version": "latest",
}
`;

exports[`URL prefix success 7`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
  "version": "dev",
}
`;

exports[`URL prefix success 8`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/456/123",
  "version": "2.1.0",
}
`;

exports[`URL suffix success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`URL suffix success 2`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
  "version": "latest",
}
`;

exports[`URL suffix success 3`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
  "version": "dev",
}
`;

exports[`URL suffix success 4`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
  "version": "2.1.0",
}
`;

exports[`namespace success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`namespace success 2`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`namespace success 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`namespace success 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo",
}
`;

exports[`namespace success 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
}
`;

exports[`namespace success 6`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
  "version": "latest",
}
`;

exports[`namespace success 7`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
  "version": "dev",
}
`;

exports[`namespace success 8`] = `
{
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
  "version": "1.2.3",
}
`;

exports[`namespace success 9`] = `
{
  "deploymentHash": "01234567",
  "deploymentIdentifier": "username/foo-bar@01234567",
  "projectIdentifier": "username/foo-bar",
  "toolPath": "/foo/bar/123",
}
`;
